### PR TITLE
Add ALTER TABLE SET PROPERTIES statement

### PR DESCRIPTION
**Pull request by pratyakshsharma**
Cherry-pick of https://github.com/trinodb/trino/pull/9401

Co-authored-by: ebyhr

## Motivation and Context
Setting table properties can help in modifying critical table properties such as the one which controls concurrency behaviour for an iceberg table.

## Impact
ALTER TABLE ... SET PROPERTIES

## Test Plan
Manual testing
Unit testing

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Add :doc:`/sql/alter-table` SET PROPERTIES statement :pr:`21495`
```

### CODE REVIEW

**Comment by Akanksha-kedia** [comment]
single commit per PR

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

**Comment by pratyakshsharma** [comment]
@Akanksha-kedia 

There are 2 separate cherry-picks which server different purposes, hence 2 different commits.
I think verifications related to existing properties and what key value pairs are going to be set can be handled at connector level. Different table formats can have different key for possible table properties.
I am adding this syntax support because ALTER TABLE SET PROPERTIES is currently needed for iceberg to solve concurrency related issues. You can refer this for more details - https://github.com/prestodb/presto/issues/21251

**Comment by pratyakshsharma** [comment]
This is ready for review now @yingsu00 @imjalpreet @agrawalreetika

**Comment by pratyakshsharma** [comment]
After the recent push, build is failing. Working on this.

**Comment by pratyakshsharma** [comment]
This is ready to be reviewed.

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseMetadata.java [hunk 1]
```diff
@@ -253,6 +253,12 @@ public void renameTable(ConnectorSession session, ConnectorTableHandle table, Sc
         clickHouseClient.renameTable(ClickHouseIdentity.from(session), tableHandle, newTableName);
     }
 
+    public void setTableProperties(ConnectorSession session, ConnectorTableHandle table, Map<String, Object> properties)
+    {
+        ClickHouseTableHandle tableHandle = (ClickHouseTableHandle) table;
+        clickHouseClient.setTableProperties(ClickHouseIdentity.from(session), tableHandle, properties);
+    }
+
```
Comment (agrawalreetika) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseMetadata.java:261]: Add @Override annotation here.
Can we add some tests to test the feature support in the connector?

**Review by agrawalreetika — CHANGES_REQUESTED** [review]

**Comment by agrawalreetika** [comment]
@pratyakshsharma Should we include `ALTER TABLE SET PROPERTIES` support commit in this PR itself? WDYT

File: presto-main/src/main/java/com/facebook/presto/metadata/AbstractPropertyManager.java [hunk 2]
```diff
@@ -125,6 +121,25 @@ public final Map<String, Object> getProperties(
...
+        return properties;
+    }
+
+    public final Map<String, Object> getProperties(
+            ConnectorId connectorId,
+            String catalog, // only use this for error messages
+            Map<String, Expression> sqlPropertyValues,
+            Session session,
+            Metadata metadata,
+            Map<NodeRef<Parameter>, Expression> parameters)
```
[the remark this answers is not in the document]
Comment (pratyakshsharma) [reply presto-main/src/main/java/com/facebook/presto/metadata/AbstractPropertyManager.java:133]: I did not understand the use case of introducing that parameter, and hence skipped it.

File: presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcClient.java [hunk 3]
```diff
@@ -73,6 +77,11 @@ PreparedStatement buildSql(ConnectorSession session, Connection connection, Jdbc
 
     void renameTable(ConnectorSession session, JdbcIdentity identity, JdbcTableHandle handle, SchemaTableName newTableName);
 
+    default void setTableProperties(ConnectorSession session, JdbcTableHandle tableHandle, Map<String, Object> properties)
```
[the remark this answers is not in the document]
Comment (pratyakshsharma) [reply presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcClient.java:80]: This is not needed. ClickhouseClient extends JdbcClient in trino, the reason they introduced it here. I ported it by mistake. Will remove, thank you for pointing this out.

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseMetadata.java (hunk 1 above)
Comment (pratyakshsharma) [reply presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseMetadata.java:261]: Clickhouse server is spawned using docker in presto tests. I skipped the test on purpose as I am not sure how do I test it. cc @yingsu00

**Comment by pratyakshsharma** [comment]
The test failures are related to the code changes from https://github.com/prestodb/presto/pull/21250. Once it gets merged into master, the failures should resolve

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

New review prompted by changes to https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/sql/alter-table.rst. Pulled branch, new local build of docs, looks good. Thanks!

**Comment by pratyakshsharma** [comment]
I see clickhouse tests are disabled in the pom.xml here - https://github.com/prestodb/presto/blob/7d1951652af763cfc0cb3421a66ff2b4f8df2d52/presto-clickhouse/pom.xml#L248.
So how do you want me to proceed with adding the clickhouse tests in this PR? @agrawalreetika @yingsu00

**Comment by pratyakshsharma** [comment]
@agrawalreetika @yingsu00 This is good to review now.

**Comment by pratyakshsharma** [comment]
@yingsu00 Raised separate PR for clickhouse as discussed internally.

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs) 

Pull branch, local doc build. Thanks!

**Comment by steveburnett** [comment]
Nit to add PR number to release note entry.
Also suggest add a link to the ALTER TABLE documentation. I tested this link example in a local doc build. 

```
== RELEASE NOTES ==

General Changes
* Add :doc:`/sql/alter-table` SET PROPERTIES statement :pr:`21495`
```

**Comment by pratyakshsharma** [comment]
@yingsu00 this is good to take a pass.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedSmokeTestBase.java [hunk 4]
```diff
@@ -1809,4 +1810,38 @@ private Session sessionForTimezone(String zoneId, boolean legacyTimestamp)
...
+    public void testUpdatingCommitRetries()
+    {
+        Session session = getSession();
+        String tableName = "test_commit_retries_update";
+        assertUpdate(session, "CREATE TABLE " + tableName + " (c1 integer, c2 varchar) WITH(commit_retries = 4)");
+        assertQuery("SELECT value FROM \"" + tableName + "$properties\" WHERE key = 'commit.retry.num-retries'", "VALUES 4");
+        assertUpdate("ALTER TABLE " + tableName + " SET PROPERTIES (commit_retries = 5)");
+        assertQuery("SELECT value FROM \"" + tableName + "$properties\" WHERE key = 'commit.retry.num-retries'", "VALUES 5");
+        assertUpdate("DROP TABLE " + tableName);
+    }
```
Comment (agrawalreetika) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedSmokeTestBase.java:1846]: Lets add few tests for `ALTER TABLE [ IF EXISTS ]` as well

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataUtil.java [hunk 5]
```diff
@@ -84,6 +91,18 @@ public static SchemaTableName toSchemaTableName(QualifiedObjectName qualifiedObj
...
+        return metadata.getCatalogHandle(session, catalogName)
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, "Catalog does not exist: " + catalogName));
+    }
+
+    public static ConnectorId getConnectorIdOrThrow(Session session, Metadata metadata, String catalogName, Statement statement, String errorMsg)
+    {
+        return metadata.getCatalogHandle(session, catalogName)
+                .orElseThrow(() -> new SemanticException(MISSING_CATALOG, statement, errorMsg, catalogName));
+    }
+
```
Comment (agrawalreetika) [inline presto-main/src/main/java/com/facebook/presto/metadata/MetadataUtil.java:105]: Since this change is unrelated to the feature, let's keep this refactoring in a separate commit?

**Review by agrawalreetika — CHANGES_REQUESTED** [review]

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedSmokeTestBase.java (hunk 4 above)
Comment (pratyakshsharma) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedSmokeTestBase.java:1847]: Done.

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataUtil.java (hunk 5 above)
Comment (pratyakshsharma) [reply presto-main/src/main/java/com/facebook/presto/metadata/MetadataUtil.java:105]: Sure.

**Comment by pratyakshsharma** [comment]
Updated release notes, thank you @steveburnett

**Review by agrawalreetika — CHANGES_REQUESTED** [review]
Overall looks good to me. 
Left few commetns around adding tests, please check

**Comment by pratyakshsharma** [comment]
@agrawalreetika @yingsu00 this is good for another pass

**Review by agrawalreetika — DISMISSED** [review]
LGTM

File: presto-main/src/main/java/com/facebook/presto/execution/SetPropertiesTask.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    public String getName()
+    {
+        return "SET PROPERTIES";
+    }
+
+    @Override
+    public ListenableFuture<?> execute(SetProperties statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, Session session, List<Expression> parameters, WarningCollector warningCollector)
+    {
+        QualifiedObjectName tableName = MetadataUtil.createQualifiedObjectName(session, statement, statement.getTableName());
+        Map<String, Expression> sqlProperties = NodeUtils.mapFromProperties(statement.getProperties());
```
Comment (hantangwangd) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPropertiesTask.java:54]: nit: use static import

File: presto-main/src/main/java/com/facebook/presto/execution/SetPropertiesTask.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+        if (statement.getType() == SetProperties.Type.TABLE) {
+            Map<String, Object> properties = metadata.getTablePropertyManager().getUserSpecifiedProperties(
+                    getConnectorIdOrThrow(session, metadata, tableName.getCatalogName()),
+                    tableName.getCatalogName(),
+                    sqlProperties,
+                    session,
+                    metadata,
+                    ParameterUtils.parameterExtractor(statement, parameters)).build();
```
Comment (hantangwangd) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPropertiesTask.java:63]: nit: static import

File: presto-main/src/main/java/com/facebook/presto/execution/RenameTableTask.java [hunk 8]
```diff
@@ -66,9 +67,7 @@ public ListenableFuture<?> execute(RenameTable statement, TransactionManager tra
         }
 
         QualifiedObjectName target = createQualifiedObjectName(session, statement, statement.getTarget());
-        if (!metadata.getCatalogHandle(session, target.getCatalogName()).isPresent()) {
-            throw new SemanticException(MISSING_CATALOG, statement, "Target catalog '%s' does not exist", target.getCatalogName());
-        }
+        MetadataUtil.getConnectorIdOrThrow(session, metadata, target.getCatalogName(), statement, targetTableCatalogError);
```
Comment (hantangwangd) [inline presto-main/src/main/java/com/facebook/presto/execution/RenameTableTask.java:70]: nit: static import

File: presto-docs/src/main/sphinx/sql/alter-table.rst [hunk 9]
```diff
@@ -14,6 +14,7 @@ Synopsis
     ALTER TABLE [ IF EXISTS ] name ADD [ CONSTRAINT constraint_name ] { PRIMARY KEY | UNIQUE } ( { column_name [, ...] } ) [ { ENABLED | DISABLED } ] [ [ NOT ] RELY ] [ [ NOT ] ENFORCED } ]
     ALTER TABLE [ IF EXISTS ] name DROP CONSTRAINT [ IF EXISTS ] constraint_name
     ALTER TABLE [ IF EXISTS ] ALTER [ COLUMN ] column_name { SET | DROP } NOT NULL
+    ALTER TABLE [ IF EXISTS ] SET PROPERTIES (property_name=value)
```
Comment (hantangwangd) [inline presto-docs/src/main/sphinx/sql/alter-table.rst:17]:
```suggestion
    ALTER TABLE [ IF EXISTS ] name SET PROPERTIES (property_name=value [, ...])
```

The properties may be more than 1, so is it better to write it in a list format?

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java [hunk 10]
```diff
@@ -186,6 +188,24 @@ public void checkCanCreateTable(ConnectorTransactionHandle transaction, Connecto
         }
     }
 
+    @Override
+    public void checkCanSetTableProperties(ConnectorTransactionHandle transactionHandle, ConnectorIdentity identity, AccessControlContext context, SchemaTableName tableName, Map<String, Object> properties)
```
Comment (hantangwangd) [inline presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:192]: Should we implement method `checkCanSetTableProperties(...)` for `SystemTableAwareAccessControl` in hive as well?

**Review by hantangwangd — COMMENTED** [review]
Thanks for this work, overall looks good to me! Some nits and little things.

**Review by ZacBlanco — COMMENTED** [review]
Mostly nits. Otherwise LGTM. One question though - is there any way to "unset" a table property through this in order to take up the default value instead? Or would a user have to manually look up the default value and set it?

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java (hunk 10 above)
Comment (pratyakshsharma) [reply presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:192]: sure.

**Comment by pratyakshsharma** [comment]
@hantangwangd @ZacBlanco @yingsu00 this is good for another pass.

**Review by hantangwangd — APPROVED** [review]
Thanks for the fix, lgtm!

**Review by yingsu00 — APPROVED** [review]
